contracts/ProcedureTable.sol

```diff
@@ -232,158 +268,428 @@ library ProcedureTable {
+        if (nTopics >= 4) {
+            if (reqTopics[3] != topic4) {
                 return false;
             }
         }
         return true;
     }
 
-    function _storeProcedure(Procedure memory p, uint192 key) internal {
-        // Get the storage address of the name/key of the procedure. In this
-        // scope "key" is the 24 byte name which is provided by the user. The
-        // procedure "p" has already been given an index (p.keyIndex) which is
-        // the offset where the name/key is stored.
-        uint248 keyPointer = _getKeyPointerByIndex(p.keyIndex);
-        // Store the name/key at this location.
-        _set(0, keyPointer, key);
+    function _storeProcedure(Procedure memory p, uint192 key, uint256[] caps) internal {
+        // Procedure List
+        // Store the the procedure key in the procedure list
+        uint256 keyPointer = _getKeyPointerByIndex(p.keyIndex);
+        _set(keyPointer, key);
 
+        // Procedure Heap
         // Get the storage address of the procedure data. This is the storage
         // key which contains all of the procedure data.
-        uint248 pPointer = _getProcedurePointerByKey(key);
-        _serialiseProcedure(p, 0, pPointer);
-    }
-
-    function _serialiseProcedure(Procedure memory p, uint8 storagePage, uint248 pPointer) internal {
-        // Store the keyIndex at this location
-        _set(storagePage, pPointer, p.keyIndex);
-        // Store the address at the loction after this (making this data 2
-        // uint256 wide).
-        _set(storagePage, pPointer + 1, uint256(p.location));
-
-        // The first value of the array is the number of capabilities
-        _set(storagePage, pPointer + 2, p.caps.length);
-        _serialiseCapArray(p, storagePage, pPointer);
-    }
-
-    function _serialiseCapArray(Procedure memory p, uint8 storagePage, uint248 pPointer) internal {
-        // n is the storage key index
-        uint248 n = 0;
-        // i is the index of the cap
-        for (uint248 i = 0; i < p.caps.length; i++) {
-            uint256 nValues = p.caps[i].values.length;
-            uint256 capSize = nValues + 1;
-            _set(storagePage, pPointer + 3 + n, capSize); n++;
-            _set(storagePage, pPointer + 3 + n, p.caps[i].capType); n++;
-            for (uint248 j = 0; j < nValues; j++) {
-                _set(storagePage, pPointer + 3 + n, p.caps[i].values[j]); n++;
+        uint256 pPointer = _getProcedurePointerByKey(key);
+        _serialiseProcedure(p, pPointer, caps);
+    }
+
+    function _serialiseProcedure(Procedure memory p, uint256 pPointer, uint256[] caps) internal {
+        // Store the address of the code contract
+        _set(pPointer + 0, uint256(p.location));
+        // Store the keyIndex
+        _set(pPointer + 1, p.keyIndex);
+        _serialiseCapArray(pPointer, caps);
+    }
+
+    function _serialiseCapArray(uint256 pPointer, uint256[] caps) internal {
+        uint192 currentProcedure = _getCurrentProcedure();
```

Comment:
I think it would be nice to have `currentProcedure` as a parameter instead of a global context.

Reply:
I know what you mean, but it is inherently a global thing, stored in storage. We could pull it from storage in the function above this, or the one above that etc. but we would need to perform the same call at some point.

Currently it is:

```javascript
registerProcedure(name, procedureAddress, caps)
_registerAnyProcedure(name, procedureAddress, caps)
procedures.insert(name, procedureAddress, caps);
_storeProcedure(key, keyIndex, location, caps);
_serialiseProcedure(pPointer, keyIndex, location, caps);
_serialiseCapArray(pPointer, caps)
```

We could extract it from storage at `procedures.insert()`, but it would just look like this:


```javascript
registerProcedure(name, procedureAddress, caps)
_registerAnyProcedure(name, procedureAddress, caps)
procedures.insert(name, procedureAddress, caps);
_storeProcedure(currentProcedure, key, keyIndex, location, caps);
_serialiseProcedure(currentProcedure, pPointer, keyIndex, location, caps);
_serialiseCapArray(currentProcedure, pPointer, caps)
```

So I don't think it adds anything, it's still an environment being pulled from storage at some point in the registration procedure. What do you think?

Reply:
I think validation should be in `Kernel.sol`, but I digress. I think it should be fine right now.

Reply:
Oh I see what you mean, though I do agree, we should definitely leave it for now. Data stack limitations made the subset and delegation checking code quite a hairy thing to implement and I think it great simplify things from an implementation perspective. Agreed.